new resource "azurerm_spring_cloud_java_deployment" and "azurerm_spring_cloud_active_deployment"

njuCZ opened the pull request:
The service team has refactored the API design, active deployment could be deleted now.

now spring cloud deployment could support different runtime: java and dotnet. We are going to split it into different resources and support java deployment first

this PR adds two resources: `azurerm_spring_cloud_java_deployment` and `azurerm_spring_cloud_active_deployment`

![image](https://user-images.githubusercontent.com/2786738/102762561-6c4b1900-43b3-11eb-914c-7c5dfe64bbec.png)

mybayern1974 commented:
@jackofallops to possibly help review this Azure Spring Cloud Deployment resource.

jackofallops commented on website/docs/r/spring_cloud_active_deployment.html.markdown:
```diff
@@ -0,0 +1,90 @@
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_spring_cloud_active_deployment"
+description: |-
+  Manage an Active Azure Spring Cloud Deployment.
+---
+
+# azurerm_spring_cloud_active_deployment
+
+Manage an Active Azure Spring Cloud Deployment.
```
```suggestion
  Manages an Active Azure Spring Cloud Deployment.
---

# azurerm_spring_cloud_active_deployment

Manages an Active Azure Spring Cloud Deployment.
```

jackofallops commented on website/docs/r/spring_cloud_active_deployment.html.markdown:
```diff
@@ -0,0 +1,90 @@
+  memory_in_gb        = 4
+  instance_count      = 2
+  jvm_options         = "-XX:+PrintGC"
+  runtime_version     = "Java_8"
```
This is the default, so perhaps set the alternative for the example?
```suggestion
  runtime_version     = "Java_11"
```

jackofallops commented on website/docs/r/spring_cloud_active_deployment.html.markdown:
```diff
@@ -0,0 +1,90 @@
+  }
+}
+
+resource "azurerm_spring_cloud_active_deployment" "test" {
```
```suggestion
resource "azurerm_spring_cloud_active_deployment" "example" {
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+  memory_in_gb        = 4
+  instance_count      = 2
+  jvm_options         = "-XX:+PrintGC"
+  runtime_version     = "Java_8"
```
As above
```suggestion
  runtime_version     = "Java_11"
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_spring_cloud_java_deployment"
+description: |-
+  Manage an Azure Spring Cloud Deployment which runtime is Java.
+---
+
+# azurerm_spring_cloud_java_deployment
+
+Manage an Azure Spring Cloud Deployment which runtime is Java.
```
```suggestion
  Manages an Azure Spring Cloud Deployment which runtime is Java.
---

# azurerm_spring_cloud_java_deployment

Manages an Azure Spring Cloud Deployment which runtime is Java.
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+  jvm_options         = "-XX:+PrintGC"
+  runtime_version     = "Java_8"
+
+  env = {
+    "Env" : "Staging"
+  }
```
Perhaps some additional entries here to help users?
```suggestion
  env = {
    "Env"    = "Staging"
    "my_var" = "someValue"
  }
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+* `spring_cloud_app_id` - (Required) Specifies the id of the Spring Cloud Application in which to create the Deployment. Changing this forces a new resource to be created.
+
+* `cpu` - (Optional) Specifies the required cpu of the Spring Cloud Deployment. Possible Values are between `1` and `4`.
```
```suggestion
* `cpu` - (Optional) Specifies the required cpu of the Spring Cloud Deployment. Possible Values are between `1` and `4`. Defaults to `1` if not specified.
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+* `cpu` - (Optional) Specifies the required cpu of the Spring Cloud Deployment. Possible Values are between `1` and `4`.
+
+* `memory_in_gb` - (Optional) Specifies the required memory size of the Spring Cloud Deployment. Possible Values are between `1` and `8`.
```
```suggestion
* `memory_in_gb` - (Optional) Specifies the required memory size of the Spring Cloud Deployment. Possible Values are between `1` and `8`. Defaults to `1` if not specified. 
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+* `memory_in_gb` - (Optional) Specifies the required memory size of the Spring Cloud Deployment. Possible Values are between `1` and `8`.
+
+* `instance_count` - (Optional) Specifies the required instance count of the Spring Cloud Deployment. Possible Values are between `1` and `500`.
```
```suggestion
* `instance_count` - (Optional) Specifies the required instance count of the Spring Cloud Deployment. Possible Values are between `1` and `500`. Defaults to `1` if not specified. 
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,97 @@
+
+* `runtime_version` - (Optional) Specifies the runtime version of the Spring Cloud Deployment. Possible Values are `Java_8` and `Java_11`. Defaults to `Java_8`.
+
+* `env` - (Optional) Specifies the environment variables of the Spring Cloud Deployment.
```
```suggestion
* `env` - (Optional) Specifies the environment variables of the Spring Cloud Deployment as a map of key-value pairs.
```

jackofallops reviewed (changes requested):
Hi @njuCZ
Thanks for this PR, and apologies for the delay in review.

Things are looking pretty good, I've left some comments and suggestions below.  I think it would be wise/cleaner to rebase against current master before addressing the comments as there have been dependency and linter updates as well as quite a few refactoring PRs.

njuCZ commented:
@jackofallops thanks for your sugestions! I have rebased the latest master branch and updated according to your suggestions. Could you have a look again?

njuCZ commented:
@jackofallops I have rebase the latest master branch, and now this PR are using stable api version. 
One thing that may need attention: in this api version, `instanceCount` field is renamed as `capacity` in the `sku` struct, but the service team still hopes to use `instance_count`, which is align to cli, powershell and portal

jackofallops commented on azurerm/internal/services/springcloud/spring_cloud_java_deployment_resource.go:
```diff
@@ -0,0 +1,248 @@
+
+	skuName := "S0"
+	skuTier := "Standard"
+	if sku := service.Sku; sku != nil {
+		if sku.Name != nil {
+			skuName = *sku.Name
+			skuTier = *sku.Tier
+		}
+	}
```
Is there a valid case where we'd get an empty response back for these values?  Feels like we should error if we don't have concrete values, rather than assume them?

njuCZ replied:
there should be no such case, I am just doing Defensive programming. I have changed it to throw error.

jackofallops commented on azurerm/internal/services/springcloud/spring_cloud_java_deployment_resource.go:
```diff
@@ -0,0 +1,248 @@
+		settings := resp.Properties.DeploymentSettings
+		d.Set("cpu", settings.CPU)
+		d.Set("memory_in_gb", settings.MemoryInGB)
+
```
```suggestion
```

jackofallops commented on azurerm/internal/services/springcloud/spring_cloud_java_deployment_resource_test.go:
```diff
@@ -0,0 +1,174 @@
+  memory_in_gb        = 4
+  instance_count      = 2
+  jvm_options         = "-XX:+PrintGC"
+  runtime_version     = "Java_8"
```
Missed this one last pass, can we have this as `Java_11` to it's a changed value in the `update` test?

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,98 @@
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_spring_cloud_java_deployment"
+description: |-
+  Manages an Azure Spring Cloud Deployment which runtime is Java.
```
```suggestion
  Manages an Azure Spring Cloud Deployment with a Java runtime.
```

jackofallops commented on website/docs/r/spring_cloud_java_deployment.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+# azurerm_spring_cloud_java_deployment
+
+Manages an Azure Spring Cloud Deployment which runtime is Java.
```
```suggestion
Manages an Azure Spring Cloud Deployment with a Java runtime.
```

jackofallops reviewed (changes requested):
Thanks @njuCZ for the updates. Just a few more items below, and I think we'll be good to test and merge.

njuCZ commented:
@jackofallops thanks for your suggestion and I have refined this PR. I have rerun the acctest in my local and all could pass. About the website-test CI fail, it seems not related with this PR.

jackofallops commented:
> @jackofallops thanks for your suggestion and I have refined this PR. I have rerun the acctest in my local and all could pass. About the website-test CI fail, it seems not related with this PR.

Thanks @njuCZ - We've addressed the failed `website-test` Action in the main branch, so that's fine.  I've run the tests, so I'll post the results here shortly and get this merged, so it should be released in v2.45.0 next week.

jackofallops commented:
Tests Passing:
![image](https://user-images.githubusercontent.com/11830746/105484462-414e3080-5ca3-11eb-9c58-66c8f5906c02.png)
